devito/passes/iet/engine.py

```diff
@@ -303,33 +307,61 @@ def _(i, mapper, counter):
         i.indexed: v.indexed,
         i._C_symbol: v._C_symbol,
     })
-    counter[base] += 1
 
 
 @abstract_object.register(CompositeObject)
-def _(i, mapper, counter):
-    base = 'o'
-    name = '%s%d' % (base, counter[base])
+def _(i, mapper, sregistry):
+    name = sregistry.make_name(prefix='o')
 
     v = i._rebuild(name)
 
     mapper[i] = v
-    counter[base] += 1
+
+
+@abstract_object.register(BlockDimension)
+def _(i, mapper, sregistry):
+    if i._depth != 2:
+        return
+
+    p = i.parent
+    pp = i.parent.parent
+
+    name0 = pp.name
+    base = sregistry.make_name(prefix=name0)
+    name1 = sregistry.make_name(prefix='%s_blk' % base)
+
+    bd = i.parent._rebuild(name1, pp)
+    d = i._rebuild(name0, bd, i._min.subs(p, bd), i._max.subs(p, bd))
+
+    mapper.update({
+        i: d,
+        i.parent: bd
+    })
```

Comment:
This looks weird. Why is that happening?

Reply:
Because if you reconstruct using `i` as "stamp", you better be sure to also replace all of its parent occurrences, or you end up with undefined symbols